directives/authDirective.js

```diff
@@ -0,0 +1,195 @@
+const { SchemaDirectiveVisitor } = require('apollo-server-express');
+const {
+  UnauthenticatedError,
+  UnauthorizedError,
+  NotFoundError,
+} = require('errors');
+const { defaultFieldResolver } = require('graphql');
+const requestContext = require('talawa-request-context');
+
+const Event = require('../models/Event');
+const Group = require('../models/Group');
+const GroupChat = require('../models/GroupChat');
+const MembershipRequest = require('../models/MembershipRequest');
+const Organization = require('../models/Organization');
+
+class AuthenticationDirective extends SchemaDirectiveVisitor {
+  visitFieldDefinition(field) {
+    const resolver = field.resolve || defaultFieldResolver;
+    field.resolve = async (root, args, context, info) => {
+      const { requires } = this.args;
+
+      if (context.expired || !context.isAuth)
+        throw new UnauthenticatedError(
+          requestContext.translate('user.notAuthenticated'),
+          'user.notAuthenticated',
+          'userAuthentication'
+        );
+
+      if (requires === 'Admin' && args.eventId) {
+        const event = await Event.findOne({ _id: args.eventId });
+        if (!event) {
+          throw new NotFoundError(
```

Comment:
@sumitra19jha @palisadoes

In every API request, we should send organizationId if it is required to check admin or not.

These 2-3 extra dB calls are too costly to check if a user is an admin or not.

What are your thoughts on this ?

Reply:
@impiyush83 Sir, One thing can be done. After fetching an event, etc. We can set the same to the request `context`. So that we need not fetch it again in the resolvers.

Reply:
Currently, I am doing the same for the organisation. This can be extended for use in events, etc.

Reply:
@aniketbiswas21   

Logically to check if a user is an admin of organization.  
`
        const isAdmin = org.admins.includes(context.userId);        if (!isAdmin) {          throw new UnauthorizedError(            requestContext.translate('user.notAuthorized'),            'user.notAuthorized',            'userAuthorization'          );        }`

Reply:
But fetching membership requests, groups, or events in authentication directives is not ideal.

Reply:
@impiyush83 So should I do the same for events, etc?

Reply:
> But fetching membership requests, groups, or events in authentication directives is not ideal. It should always be in resolvers where business logic resides.

Ah, understood.

Reply:
There is no need for these 2-3 DB calls at all. Our purpose is not at all aimed towards anything related to the event at all.

But since we are having a conversation to reduce the DB query then in these cases, small information can be encoded in JWT itself, and to increase the security we can use bcrypt.

Reply:
But there is no need for these 2-3 calls